lapce-core/src/language.rs

```diff
@@ -4,179 +4,388 @@ use tree_sitter::{Parser, TreeCursor};
+// the crate named as tree-sitter-foo:
+//
+// 1. Add an optional dependency on tree-sitter-foo in this crate.
+//
+//    [dependencies]
+//    # ...
+//    tree-sitter-foo = { version = "1", optional = true }
+//
+// 2. Add a new feature, say "lang-foo", to this crate to use this dependency.
+//    Also add "lang-foo" to the "all-languages" feature (see
+//    lapce-core/Cargo.toml).
+//
+//    [features]
+//    # ...
+//    lang-foo = "dep:tree-sitter-foo"
+//
+// 3. Add a new variant to `LapceLanguage`, say Foo, following the existing
+//    variants, guard the new variant with the new feature.
+//
+//    pub enum LapceLanguage {
+//         // ...
+//         #[cfg(feature = "lang-foo")]
+//         Foo,
+//    }
+//
+// 4. Add a new element in the LANGUAGES array, guard the new element with the
+//    new feature.
+//
+//    const LANGUAGES: &[Settings] = &[
+//        // ...
+//        #[cfg(feature = "lang-foo")]
+//        Setting{
+//            id: LapceLanguage::Foo,
+//            language: tree_sitter_foo::language,
+//            highlight: tree_sitter_foo::HIGHLIGHT_QUERY,
+//            comment: "//",
+//            indent: "    ",
+//            code_lens: (&[/* ... */], &[/* ... */]),
+//        },
+//    ];
+//
+// 5. Add a new match arm in `LapceLanguage::from_path`, guard the new arm with
+//    the new feature.
+//
+//    Some(match extension.as_str() {
+//        // ...
+//        #[cfg(feature = "lang-foo")]
+//        "foo" => LapceLanguage::Foo,
+//        _ => return None,
+//    })
+//
+// 6. Add a new feature, say "lang-foo", to the lapce-ui crate (see
+//    lapce-ui/Cargo.toml).
+//
+//    [features]
+//    # ...
+//    lang-foo = "lapce-core/lang-foo"
+//
+
+// Use these lists when a language does not have specific settings for "code
+// lens".
+#[allow(dead_code)]
 const DEFAULT_CODE_LENS_LIST: &[&str] = &["source_file"];
+#[allow(dead_code)]
 const DEFAULT_CODE_LENS_IGNORE_LIST: &[&str] = &["source_file"];
-const RUST_CODE_LENS_LIST: &[&str] =
-    &["source_file", "impl_item", "trait_item", "declaration_list"];
-const RUST_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "use_declaration", "line_comment"];
-const GO_CODE_LENS_LIST: &[&str] = &[
-    "source_file",
-    "type_declaration",
-    "type_spec",
-    "interface_type",
-    "method_spec_list",
-];
-const GO_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "comment", "line_comment"];
-const PYTHON_CODE_LENS_LIST: &[&str] = &[
-    "source_file",
-    "module",
-    "class_definition",
-    "class",
-    "identifier",
-    "decorated_definition",
-    "block",
-];
-const PYTHON_CODE_LENS_IGNORE_LIST: &[&str] =
-    &["source_file", "import_statement", "import_from_statement"];
-const JAVASCRIPT_CODE_LENS_LIST: &[&str] = &["source_file", "program"];
-const JAVASCRIPT_CODE_LENS_IGNORE_LIST: &[&str] = &["source_file"];
 
-#[derive(Eq, PartialEq, Hash, Copy, Clone, Debug)]
+struct Setting {
```

Comment:
I think this struct should have the file extensions as well. I'm also not sure about the name (these are not settings, but items in a pre-defined set).

Reply:
> I think this struct should have the file extensions as well. I'm also not sure about the name (these are not settings, but items in a pre-defined set).

For the struct name, how about `SyntaxProperties`?

I actually thought about having the file extensions in this struct. I give up because the `from_path()` method will then probably not be able to use pattern match any more.

Do you think a linear search in the `LANGUAGES` array will be acceptable? It will not be in any tight loops.

Reply:
I don't think a linear search is a problem at this scale. I'm not entirely convinced that pattern matching is not linear in this case, anyway.

`SyntaxProperties` sounds fine to me.

Reply:
> I don't think a linear search is a problem at this scale. I'm not entirely convinced that pattern matching is not linear in this case, anyway.

I will try to implement the linear search then. That will cut down the need for cfg feature gates, too.

> `SyntaxProperties` sounds fine to me.

Ok.